File: csharp/day2/Program.cs
using System;
using Utils;
using System.Collections.Generic;
using System.Linq;

namespace adventofcode2020
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Day 2 of AoC");
            IEnumerable<Tuple<string, int>> data = null;
            Utils.Utils.MeasureActionTime("Load", () =>
            {
                string input = args[0];
                Console.WriteLine($"Loading data...");
                data = LoadData(input);
                Console.WriteLine($"Length -> {data.Count()}");
            });
            Utils.Utils.MeasureActionTime("Part 1", () =>
            {
                var res = DoMagic(data);
                Console.WriteLine($"Result -> {res}");
            });
            Utils.Utils.MeasureActionTime("Part 2", () =>
            {
                var res = DoMagic2(data);
                Console.WriteLine($"Result -> {res}");
            });
            Console.WriteLine("END");
        }

        static IEnumerable<Tuple<string, int>> LoadData(string filename)
        {                
            var raw = Utils.Utils.ReadAllLines(filename);
            foreach (var item in raw)
            {
                var l = item.Split(' ');
                yield return new Tuple<string, int>(l[0], Int32.Parse(l[1]));
            }
        }

        static int DoMagic(IEnumerable<Tuple<string, int>> input)
        {        
            var x = 0;
            var y = 0;
            foreach (var item in input)
            {
                switch (item.Item1)
                {
                    case "forward":
                        x += item.Item2;
                        break;
                    case "down":
                        y += item.Item2;
                        break;
                    case "up":
                        y -=item.Item2;
                        break;
                }
            }

            return x * y;
        }

        static int DoMagic2(IEnumerable<Tuple<string, int>> input)
        {
            var x = 0;
            var y = 0;
            var aim = 0;

            foreach (var item in input)
            {
                switch (item.Item1)
                {
                    case "forward":
                        x += item.Item2;
                        y = y + item.Item2*aim; 
                        break;
                    case "down":
                        aim += item.Item2;
                        break;
                    case "up":
                        aim -=item.Item2;
                        break;
                }
            }

            return x * y;
        }
    }
}

File: python/day2/program.py
import sys
import os
import math

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append(os.path.dirname(SCRIPT_DIR + "/../"))

from utils import utils

def main():
    input = loadData()
    part1(input)
    part2(input)    

def loadData():
    input = utils.readAllLines()
    res = []
    for i in range(0, len(input)):
        cmd = input[i].split()
        res.append((cmd[0],int(cmd[1])))
    return res

def part1(input):
    x = 0
    y = 0
    for cmd in input:
        match cmd[0]:
            case "forward":
                x += cmd[1]
            case "up":
                y -= cmd[1]
            case "down":
                y += cmd[1]

    print("AoC part 1: " + str(x*y))

def part2(input):
    x = 0
    y = 0
    aim = 0
    for cmd in input:
        match cmd[0]:
            case "forward":
                x += cmd[1]
                y += cmd[1]*aim
            case "up":
                aim -= cmd[1]
            case "down":
                aim += cmd[1]

    print("AoC part 1: " + str(x*y))

if __name__ == "__main__":
    main()